PR: https://github.com/odoo/odoo/pull/222014

From: f95c3fdf68ce106ad98aacf84f1db00d9d1e9de7
From: Ethan Vincent
Date: 2025-08-28 08:53:40

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'stock.valuation.layer'

Total Changes: 85

[FIX] {mrp,stock}_account, purchase_mrp: unbuild cost difference correction

**Current behavior:**
Since 84dda96814, we valuate an unbuild operation by attempting
to match the ensuing OUT layer with the IN layer from the
original MO.

The point being to eliminate a potential valuation imbalance for
the manufactured product (might affect cost, etc.).

**New behavior**
Don't attempt to match an unbuild valuation layer with the
original MO valuation layer. If there is some cost difference
between build time and unbuild time, make a corrective journal
entry for it.

**Issue with current behavior**
The following sequence:
1. Create Product A with average costing, real-time valuation
2. Create 2 components, avg costing, real-time val
3. Create a BoM for Product A with the components
4. Manufacture 3 units of Product A with different component
quantities (can set flexible consumption on BoM)
5. Unbuild the first manufactured unit

Results in Product A's cost not matching the expected average
cost according to the valuation layers.

**Cause of the issue:**
Unbuilding the first MO created an out move at the original
"build time" cost, but since we've built 2 additional qty, that
original cost is not the current average cost- thus the
"theoretically current" standard price of the product (sum of
layers value divided by remaining qty) is no longer the value we
see on the product form.

**Fix:**
Don't try and match the valuation layers. Aside from issues such
as the one described above, it might not actually make
functional sense to do so (e.g., FIFO isn't actually adhering to
first-in-first-out if we're preferring the original IN layer for
an unbuild valuation).

Instead we make a journal entry with any excess cost of
production (it was actually a solution proposed in the
discussion on 84dda96814 in the first place).

Additionally, now that we aren't doing this mapping, we can
revert the non-test difference of 3a69456a29, as this commit was
fixing an issue introduced by the change which added the
mapping.

opw-4877597

closes odoo/odoo#222014

X-original-commit: fb30cde9a320c245cf1321c9dc2ea2e67a53d0a0
Signed-off-by: William Henrotin (whe) <whe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp_account/models/__init__.py
+++ b/addons/mrp_account/models/__init__.py
@@ -7,4 +7,3 @@ from . import mrp_routing
 from . import product
 from . import stock_move
 from . import account_move
-from . import stock_valuation_layer

--- a/addons/mrp_account/models/stock_move.py
+++ b/addons/mrp_account/models/stock_move.py
@@ -3,7 +3,7 @@
 
 from collections import defaultdict
 
-from odoo import models
+from odoo import _, Command, fields, models
 
 
 class StockMove(models.Model):
@@ -41,38 +41,42 @@ class StockMove(models.Model):
         self.ensure_one()
         return self.location_dest_id.usage == 'production' and self.location_id._should_be_valued()
 
-    def _get_out_svl_vals(self, forced_quantity):
-        unbuild_moves = self.filtered('unbuild_id')
-        # 'real cost' of finished product moves @ build time
-        price_unit_map = {
-            move.id: (
-                (move.unbuild_id.mo_id.move_finished_ids |
-                move.unbuild_id.mo_id.move_raw_ids).stock_valuation_layer_ids.filtered(
-                    lambda svl: svl.product_id == move.product_id
-                )[0].unit_cost,
-                move.company_id.currency_id.round,
-            )
-            for move in unbuild_moves.sudo()
-            if move.product_id.cost_method != 'standard' and
-            move.unbuild_id.mo_id.move_finished_ids.stock_valuation_layer_ids
-        }
-        svl_vals_list = super()._get_out_svl_vals(forced_quantity)
-        if price_unit_map:
-            for svl_vals in svl_vals_list:
-                if (move_id := svl_vals['stock_move_id']) in price_unit_map:
-                    unit_cost = price_unit_map[move_id][0]
-                    svl_vals.update({
-                        'unit_cost': unit_cost,
-                        'value': price_unit_map[move_id][1](unit_cost * svl_vals['quantity']),
-                    })
-        return svl_vals_list
-
     def _create_out_svl(self, forced_quantity=None):
-        product_unbuild_map = defaultdict(self.env['mrp.unbuild'].browse)
-        for move in self:
-            if move.unbuild_id:
-                product_unbuild_map[move.product_id] |= move.unbuild_id
-        return super(StockMove, self.with_context(product_unbuild_map=product_unbuild_map))._create_out_svl(forced_quantity)
+        svls = super()._create_out_svl(forced_quantity)
+        unbuild_svls = svls.filtered('stock_move_id.unbuild_id')
+        unbuild_cost_correction_move_list = list()
+        for svl in unbuild_svls:
+            build_time_unit_cost = svl.stock_move_id.unbuild_id.mo_id.move_finished_ids.filtered(
+                lambda m: m.product_id == svl.product_id
+            ).stock_valuation_layer_ids.unit_cost
+            unbuild_difference = svl.unit_cost - build_time_unit_cost
+            if svl.product_id.valuation == 'real_time' and not svl.currency_id.is_zero(unbuild_difference):
+                product_accounts = svl.product_id.product_tmpl_id.get_product_accounts()
+                valuation_account, production_account = (
+                    product_accounts['stock_valuation'],
+                    product_accounts['production'],
+                )
+                desc = _('%s - Unbuild Cost Difference', svl.stock_move_id.unbuild_id.name)
+                unbuild_cost_correction_move_list.append({
+                    'journal_id': product_accounts['stock_journal'].id,
+                    'date': fields.Date.context_today(self),
+                    'ref': desc,
+                    'move_type': 'entry',
+                    'line_ids': [Command.create({
+                        'name': desc,
+                        'ref': desc,
+                        'account_id': account.id,
+                        'balance': balance,
+                        'product_id': svl.product_id.id,
+                    }) for account, balance in (
+                        (valuation_account, unbuild_difference),
+                        (production_account, -unbuild_difference),
+                    )],
+                })
+        if unbuild_cost_correction_move_list:
+            unbuild_cost_correction_moves = self.env['account.move'].sudo().create(unbuild_cost_correction_move_list)
+            unbuild_cost_correction_moves._post()
+        return svls
 
     def _get_all_related_sm(self, product):
         moves = super()._get_all_related_sm(product)

--- a/addons/mrp_account/models/stock_valuation_layer.py
+++ b/None
@@ -1,16 +0,0 @@
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models
-
-
-class StockValuationLayer(models.Model):
-    _inherit = 'stock.valuation.layer'
-
-    def _candidate_sort_key(self):
-        self.ensure_one()
-        res = super()._candidate_sort_key()
-        if self.product_id in self.env.context.get('product_unbuild_map', ()):
-            unbuild = self.env.context['product_unbuild_map'][self.product_id]
-            # Give priority to the SVL that produced `self.product_id`
-            res += (self.stock_move_id.id not in unbuild.mo_id.move_finished_ids.ids,)
-        return res
